libs/depgraph/depgraph_api.go

```diff
@@ -0,0 +1,608 @@
+// Copyright (c) 2022 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package depgraph
+
+// Graph is a dependency graph.
+// The main use-case is to represent configuration items (network interfaces, routes,
+// volumes, etc.) or any managed stateful objects (incl. processes, containers, files,
+// etc.) as graph nodes and their dependencies as directed graph edges.
+// For more information please see README.md.
+type Graph interface {
+	GraphR
+
+	// SetDescription updates description assigned to the (sub)graph.
+	SetDescription(string)
+
+	// PutNode adds or moves (and updates) node into this (sub)graph.
+	PutNode(node *Node)
+	// DelNode deletes an existing node from this (sub)graph.
+	// Returns true if the node existed and was actually deleted.
+	DelNode(NodeID) bool
+
+	// PutSubGraph adds a new subgraph into this graph or updates an existing
+	// subgraph. This refers to a direct child of this graph, cannot add/update
+	// a nested subgraphs.
+	PutSubGraph(Graph)
+	// DelSubGraph deletes existing subgraph. This refers to a direct child of this
+	// graph, cannot delete a nested subgraph.
+	// Returns true if the subgraph existed and was actually deleted.
+	// It is an error to try to use a subgraph after it was deleted (can't be used
+	// even as a separate graph anymore).
+	DelSubGraph(name string) bool
+	// EditSubGraph elevates read-only subgraph handle to read-write access.
+	// Panics if the given graph is not actually a subgraph (direct or nested)
+	// of this graph.
+	EditSubGraph(GraphR) Graph
+	// EditParentGraph returns read-write handle to a (direct) parent graph
+	// of this subgraph.
+	// Return nil if the graph is not a subgraph.
+	EditParentGraph() Graph
```

Comment:
Two-way graph. Does it add complexity? Do we need this, is it ever used?

Reply:
It doesn't increase the complexity.
It is used only for a special case - If you do:
```
// Reconcile just a subset of the current state by selecting a subgraph.
reconciler.Reconcile(subGraph, nil)  // intended state is nil
```
Then the semantics (which we can of course define as we wish) is currently to un-configure (Delete) everything that is inside the subgraph and to remove the subGraph from the current state. Just for that the reconciler needs to get handle to the parent.

Reply:
Got it. Makes sense. And is long as the addition/removal of the child->parent link is handled by the layer underneath - nothing the user needs to do to manage it - then it is fine.